metaflow/plugins/pypi/pip.py

```diff
@@ -89,21 +109,76 @@ def download(self, id_, packages, python, platform):
         # download packages only if they haven't ever been downloaded before
         if os.path.isfile(metadata_file):
             return
+
         metadata = {}
+        custom_index_url, extra_index_urls = self.indices(prefix)
+
+        # build wheels if needed
+        with ThreadPoolExecutor() as executor:
+
+            def _build(key, package):
+                dest = "{prefix}/.pip/built_wheels/{key}".format(prefix=prefix, key=key)
+                cmd = [
+                    "wheel",
+                    "--no-deps",
+                    "--progress-bar=off",
+                    "--wheel-dir=%s" % dest,
+                    "--quiet",
+                    *(["--index-url", custom_index_url] if custom_index_url else []),
+                    *(
+                        chain.from_iterable(
+                            product(["--extra-index-url"], set(extra_index_urls))
+                        )
+                    ),
+                    package["url"],
+                ]
+                self._call(prefix, cmd)
+                return package, dest
+
+            results = list(
+                executor.map(
+                    lambda x: _build(*x),
+                    enumerate(
+                        package for package in packages if package["require_build"]
+                    ),
+                )
+            )
+            os.makedirs("%s/.pip/wheels" % prefix, exist_ok=True)
+
+            for package, path in results:
+                (wheel,) = [
+                    f
+                    for f in os.listdir(path)
+                    if os.path.isfile(os.path.join(path, f)) and f.endswith(".whl")
+                ]
+                if (
+                    len(set(pip_tags(python, platform)).intersection(wheel_tags(wheel)))
+                    == 0
+                ):
+                    raise PipException(
+                        "The built wheel %s is not supported for %s with Python %s"
+                        % (wheel, platform, python)
+                    )
+                target = "{prefix}/.pip/wheels/{wheel}".format(
+                    prefix=prefix, wheel=wheel
+                )
+                shutil.move(os.path.join(path, wheel), target)
+                metadata["{url}".format(**package)] = target
```

Comment:
slight note on comparing this to #1559 , keeping a separate `BUILD_METADATA` file has the benefit of being able to skip only sources requiring building if they've been built already for the environment. Consider the case:
- user has both pypi and git sources
- git sources get built first, successfully
- pypi sources fail to download due to network issues
- environment creation fails

retrying this would require rebuilding the git sources as well unnecessarily?

Reply:
yes, that's a very valid edge case but comes at the expense of managing state across multiple files. hopefully re-building a pure-python package should be rather quick but if that isn't the case, we can always split the file into two.